File: MTGSearch/src/test/java/com/dbottillo/mtgsearchfree/interactors/PlayerInteractorImplTest.java
package com.dbottillo.mtgsearchfree.interactors;

import com.dbottillo.mtgsearchfree.model.Player;
import com.dbottillo.mtgsearchfree.model.storage.PlayersStorage;
import com.dbottillo.mtgsearchfree.util.Logger;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;

import java.util.Arrays;
import java.util.List;

import io.reactivex.observers.TestObserver;

import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class PlayerInteractorImplTest {

    @Rule
    public MockitoRule mockitoRule = MockitoJUnit.rule();

    @Mock
    PlayersStorage storage;
    @Mock
    Player player;
    @Mock
    Logger logger;

    private List<Player> players = Arrays.asList(new Player(1, "Jace"), new Player(2, "Liliana"));
    private List<Player> toEditPlayers = Arrays.asList(new Player(3, "Chandra"), new Player(4, "Sorin"));

    private PlayerInteractor underTest;

    @Before
    public void setup() {
        when(storage.load()).thenReturn(players);
        when(storage.addPlayer(player)).thenReturn(players);
        when(storage.removePlayer(player)).thenReturn(players);
        when(storage.editPlayer(player)).thenReturn(players);
        when(storage.editPlayers(toEditPlayers)).thenReturn(players);
        underTest = new PlayerInteractorImpl(storage, logger);
    }

    @Test
    public void testLoad() {
        TestObserver<List<Player>> testSubscriber = new TestObserver<>();
        underTest.load().subscribe(testSubscriber);
        testSubscriber.assertNoErrors();
        testSubscriber.assertValue(players);
        verify(storage).load();
    }

    @Test
    public void testAddPlayer() {
        TestObserver<List<Player>> testSubscriber = new TestObserver<>();
        underTest.addPlayer(player).subscribe(testSubscriber);
        testSubscriber.assertNoErrors();
        testSubscriber.assertValue(players);
        verify(storage).addPlayer(player);
    }

    @Test
    public void testEditPlayer() {
        TestObserver<List<Player>> testSubscriber = new TestObserver<>();
        underTest.editPlayer(player).subscribe(testSubscriber);
        testSubscriber.assertNoErrors();
        testSubscriber.assertValue(players);
        verify(storage).editPlayer(player);
    }

    @Test
    public void testEditPlayers() {
        TestObserver<List<Player>> testSubscriber = new TestObserver<>();
        underTest.editPlayers(toEditPlayers).subscribe(testSubscriber);
        testSubscriber.assertNoErrors();
        testSubscriber.assertValue(players);
        verify(storage).editPlayers(toEditPlayers);
    }

    @Test
    public void testRemovePlayer() {
        TestObserver<List<Player>> testSubscriber = new TestObserver<>();
        underTest.removePlayer(player).subscribe(testSubscriber);
        testSubscriber.assertNoErrors();
        testSubscriber.assertValue(players);
        verify(storage).removePlayer(player);
    }
}

File: MTGSearch/src/test/kotlin/com/dbottillo/mtgsearchfree/interactors/PlayerInteractorImplTest.kt
package com.dbottillo.mtgsearchfree.interactors

import com.dbottillo.mtgsearchfree.model.Player
import com.dbottillo.mtgsearchfree.storage.PlayersStorage
import com.dbottillo.mtgsearchfree.util.Logger
import com.nhaarman.mockito_kotlin.verify
import com.nhaarman.mockito_kotlin.verifyNoMoreInteractions
import com.nhaarman.mockito_kotlin.whenever
import io.reactivex.observers.TestObserver
import io.reactivex.schedulers.Schedulers
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.mockito.Mock
import org.mockito.junit.MockitoJUnit

class PlayerInteractorImplTest {

    @Rule @JvmField var mockitoRule = MockitoJUnit.rule()!!

    @Mock internal lateinit var storage: PlayersStorage
    @Mock internal lateinit var player: Player
    @Mock internal lateinit var logger: Logger
    @Mock lateinit var players: List<Player>
    @Mock lateinit var schedulerProvider: SchedulerProvider

    private lateinit var underTest: PlayerInteractor

    @Before
    fun setup() {
        whenever(schedulerProvider.io()).thenReturn(Schedulers.trampoline())
        whenever(schedulerProvider.ui()).thenReturn(Schedulers.trampoline())
        underTest = PlayerInteractorImpl(storage, schedulerProvider, logger)
    }

    @Test
    fun `load players, should call storage and return observable`() {
        whenever(storage.load()).thenReturn(players)
        val testSubscriber = TestObserver<List<Player>>()

        val result = underTest.load()

        result.subscribe(testSubscriber)
        testSubscriber.assertNoErrors()
        testSubscriber.assertValue(players)
        verify(storage).load()
        verify(schedulerProvider).io()
        verify(schedulerProvider).ui()
        verifyNoMoreInteractions(storage, schedulerProvider)
    }

    @Test
    fun `add player, should call storage and return observable`() {
        whenever(storage.addPlayer()).thenReturn(players)
        val testSubscriber = TestObserver<List<Player>>()

        underTest.addPlayer().subscribe(testSubscriber)

        testSubscriber.assertNoErrors()
        testSubscriber.assertValue(players)
        verify(storage).addPlayer()
        verify(schedulerProvider).io()
        verify(schedulerProvider).ui()
        verifyNoMoreInteractions(storage, schedulerProvider)
    }

    @Test
    fun `edit player, should call storage and return observable`() {
        whenever(storage.editPlayer(player)).thenReturn(players)
        val testSubscriber = TestObserver<List<Player>>()

        underTest.editPlayer(player).subscribe(testSubscriber)

        testSubscriber.assertNoErrors()
        testSubscriber.assertValue(players)
        verify(storage).editPlayer(player)
        verify(schedulerProvider).io()
        verify(schedulerProvider).ui()
        verifyNoMoreInteractions(storage, schedulerProvider)
    }

    @Test
    fun `edit players, should call storage and return observable`() {
        whenever(storage.editPlayers(players)).thenReturn(players)
        val testSubscriber = TestObserver<List<Player>>()

        underTest.editPlayers(players).subscribe(testSubscriber)

        testSubscriber.assertNoErrors()
        testSubscriber.assertValue(players)
        verify(storage).editPlayers(players)
        verify(schedulerProvider).io()
        verify(schedulerProvider).ui()
        verifyNoMoreInteractions(storage, schedulerProvider)
    }

    @Test
    fun `remove player, should call storage and return observable`() {
        whenever(storage.removePlayer(player)).thenReturn(players)
        val testSubscriber = TestObserver<List<Player>>()

        underTest.removePlayer(player).subscribe(testSubscriber)

        testSubscriber.assertNoErrors()
        testSubscriber.assertValue(players)
        verify(storage).removePlayer(player)
        verify(schedulerProvider).io()
        verify(schedulerProvider).ui()
        verifyNoMoreInteractions(storage, schedulerProvider)
    }
}